src/v/pandaproxy/schema_registry/storage.h

```diff
@@ -372,6 +372,111 @@ class schema_value_handler final : public json::base_handler<Encoding> {
+    static constexpr topic_key_type keytype{topic_key_type::config};
+    std::optional<subject> sub;
+    topic_key_magic magic{0};
+
+    friend bool operator==(const config_key& lhs, const config_key& rhs) {
+        return lhs.sub == rhs.sub && lhs.magic == rhs.magic;
+    }
+
+    friend std::ostream& operator<<(std::ostream& os, const config_key& v) {
+        fmt::print(
+          os,
+          "keytype: {}, subject: {}, magic: {}",
+          to_string_view(v.keytype),
+          v.sub.value_or(invalid_subject),
+          v.magic);
+        return os;
+    }
+};
+
+inline void rjson_serialize(
+  rapidjson::Writer<rapidjson::StringBuffer>& w,
+  const schema_registry::config_key& key) {
+    w.StartObject();
+    w.Key("keytype");
+    ::json::rjson_serialize(w, to_string_view(key.keytype));
+    w.Key("subject");
+    if (key.sub) {
+        ::json::rjson_serialize(w, key.sub.value());
+    } else {
+        w.Null();
+    }
+    w.Key("magic");
+    ::json::rjson_serialize(w, key.magic);
+    w.EndObject();
+}
+
+template<typename Encoding = rapidjson::UTF8<>>
+class config_key_handler : public json::base_handler<Encoding> {
+    enum class state {
+        empty = 0,
+        object,
+        keytype,
+        subject,
+        magic,
+    };
+    state _state = state::empty;
+
+public:
+    using Ch = typename json::base_handler<Encoding>::Ch;
+    using rjson_parse_result = config_key;
+    rjson_parse_result result;
+
+    config_key_handler()
+      : json::base_handler<Encoding>{json::serialization_format::none} {}
+
+    bool Key(const Ch* str, rapidjson::SizeType len, bool) {
+        auto sv = std::string_view{str, len};
+        std::optional<state> s{string_switch<std::optional<state>>(sv)
+                                 .match("keytype", state::keytype)
+                                 .match("subject", state::subject)
+                                 .match("magic", state::magic)
+                                 .default_match(std::nullopt)};
+        return s.has_value() && std::exchange(_state, *s) == state::object;
+    }
+
+    bool Uint(int i) {
+        result.magic = topic_key_magic{i};
+        return std::exchange(_state, state::object) == state::magic;
+    }
+
+    bool String(const Ch* str, rapidjson::SizeType len, bool) {
+        auto sv = std::string_view{str, len};
+        switch (_state) {
+        case state::keytype: {
+            auto kt = from_string_view<topic_key_type>(sv);
+            _state = state::object;
+            return kt.has_value() && result.keytype == *kt;
+        }
+        case state::subject: {
+            result.sub = subject{ss::sstring{sv}};
+            _state = state::object;
+            return true;
+        }
+        case state::empty:
+        case state::object:
+        case state::magic:
+            return false;
+        }
+        return false;
+    }
+
+    bool Null() {
+        return std::exchange(_state, state::object) == state::subject;
```

Comment:
how does this make sense for `null`?  a little comment would be helpful. 

checking for `null` changes the internal state to an object and checks if subject. seems like advanced-foo to me.

Reply:
This is not checking for null, it's handling a null json value.

The subject, and only the subject, is nullable.

The full expression, which "correctly" doesn't transition state on failure, is a little less terse, and a little more branchy:
```cpp
switch (_state) {
    case state::subject) {
        result.subject = std::nullopt; // already the default
        _state = state::object; // subject handled, transition to object
        return true; // all good
    }
    case state::empty:
    case state::object:
    case state::keytype:
    case state::magic:
        return false; // unexpected null
}
return false; // unexpected state, g++
```
One day I'll figure out how to code-gen this with templates